### PR TITLE
Feat/theclog

### PR DESCRIPTION
**Pull request by lnist**
Introduces
* TheClog.t.sol to test out of gas and stack too deep clogs due to failing offers.
* MaxDepth in global to configure a maxDepth (Consider name, and should it be compile time constant?)
* Accounts for needed gas for posthooks and verifies that we have enough for execute+posthook or else ends market order
* Factors out the end of market order case into its own function for readability (this is the final commit)

Since this manipulates the same code as feat/simplify-price-match-flow that branch was merged in, do not consider those changes part of this PR. The final commit and the merge makes this hard to review. The most important changes are:
1) Initialization of maxDepth and gasreqLeft in generalMarketOrder
2) Using the values and updating them in internalMarketOrder

The change unclogs at the cost of some accounting, currently taking up some bits for dynamic configuration, and two additional uints in the mor structure.

### CODE REVIEW

File: src/MgvOfferTaking.sol
Comment (espendk) [inline src/MgvOfferTaking.sol]: This is an over-approximation, right?

File: src/MgvOfferTaking.sol
Comment (espendk) [inline src/MgvOfferTaking.sol]: Here we haven't used all of the `offer_gasbase` gas set aside in line 209 since we're not done processing the offer yet. So I suspect this is an under approximation.

But maybe this is compensated for in line 156 where we set `offer_gasbase` aside for ending the order?

File: src/MgvOfferTaking.sol
Comment (espendk) [inline src/MgvOfferTaking.sol]: I like how this simplifies this method and makes it clear what the end-conditions are and how they are handled.

Comments need to be updated, but we can do that when/if we decide to move forward with this.

File: src/MgvGovernable.sol
Comment (espendk) [inline src/MgvGovernable.sol]: Two comments:

1. Perhaps we should name this parameter in a more semantic way. Eg `maxOfferTakeCount`, `maxTakenOffers` or smth. Or perhaps we should say what 'depth' this is referring to.

**Review by espendk — COMMENTED** [review]
I like this. It's annoying that we have to do it, but as you demonstrate, The Clog is real and IMO we should address it.

Have you measured the impact of gas usage by this change?

File: src/MgvGovernable.sol
Comment (lnist) [reply src/MgvGovernable.sol]: Agree - @adhusson : Do you have any preferred name base on Espen's suggestions?

File: src/MgvOfferTaking.sol
Comment (lnist) [reply src/MgvOfferTaking.sol]: You're on the right track. The reservation in 156 is needed because 258 is an underapproximation. It goes like this when taking two offers:
1. Line 258 reached first time, we have at least `offer_gasbase+gasreq1-gasused1` gas left for rounding off the market order due to what is set aside in 156 and the accounting in internalMarketOrder. It is larger due to line 209 also conditioning on additional offer_gasbase.
2. Line 258 reached the second time we have gasreqLeft be `offer_gasbase +gasreq1-gasused1+gasreq2-gasused2`, however, note that we will not execute offer2 unless gasleft() can also cover an additional offer_gasbase when reaching line 209. 

So we will never start executing an offer unless we also have `offer_gasbase` left on top of the `gasreq` . This should always be sufficient to complete a market order of 1 offer. For many offers this code relies on offer_gasbase being able to cover for the overhead of unwrapping the stack and doing the necessary updates.
Alternatively we should add some constant overhead for each offer on line 258 or let the reservation on line 156 be another value which we know is big enough to cover the unwrap of stack.

Note: gasreq for one offer is the worst case for that offer, so with a set of offers each recursion will set aside too much in case worst case is not hit, so on average a bit extra is set aside for each offer.

File: src/MgvOfferTaking.sol
Comment (lnist) [reply src/MgvOfferTaking.sol]: See other comment.

File: src/MgvOfferTaking.sol
Comment (adhusson) [reply src/MgvOfferTaking.sol]: Later when we add comments we should remember to include all this information.

File: src/MgvGovernable.sol
Comment (adhusson) [reply src/MgvGovernable.sol]: I would say `maxRecursionDepth`

**Comment by adhusson** [comment]
Great PR :) I would like to fully review&merge this only once the diff has become "smaller" (ie after we merged @espendk's changes to execution flow) , is this OK?

File: src/MgvGovernable.sol
Comment (lnist) [reply src/MgvGovernable.sol]: Alright, I'll make the change after you do the thorough review.

**Comment by lnist** [comment]
@adhusson : Definitely fine - hopefully the other is is in soon.

**Comment by espendk** [comment]
Umm, if we need the other change in first, either we have to wait for @jkrivine to have time to review it (including getting acquainted with the new ticks code) or decide it's sufficient that the three of us have approved it 🙂

File: src/MgvGovernable.sol
Comment (lnist) [reply src/MgvGovernable.sol]: The change has been made.

File: src/MgvOfferTaking.sol
Comment (adhusson) [inline src/MgvOfferTaking.sol]: Why moving to an additional function? It uses additional gas & reduces the total available stack space a little bit (edit: I understood, see other comment)

File: test/core/TheClog.t.sol
Comment (adhusson) [inline test/core/TheClog.t.sol]: `testFail` will be [soon](https://github.com/foundry-rs/foundry/pull/4969) deprecated/discouraged by foundry so I think we could move this test and the one below to a "normal" test?

File: src/MgvOfferTaking.sol
Comment (adhusson) [reply src/MgvOfferTaking.sol]: @lnist line numbers have gone stale!(?) I could use a refresh 😅

File: src/MgvOfferTaking.sol
Comment (adhusson) [inline src/MgvOfferTaking.sol]: Ok I understand why the function `endInternalMarketOrder` was created!

Since moving to ticks, we have 72 bits available in `offer`. I think moving `gasreq` from `offerDetail` to `offer` make sense. It would reduce gas use in some situations, and mainly here it may remove one call to `endInternalMarketOrder` because we could just add `|| gasleft() < ...` to the big `if` at the beginning of `internalMarketOrder`. With the proviso that updating `mor.gasreqLeft` early is not a problem.

By removing one call to it we would then be able to inline it.

File: test/core/TakerOperations.t.sol
Comment (adhusson) [inline test/core/TakerOperations.t.sol:774]: How can that revert still happen? I think I'm missing something

**Review by adhusson — CHANGES_REQUESTED** [review]
Extremely nice commit-by-commit decomposition :)

I _think_ this could use some additional tests re: checking what Mangrove does when the max stack is reached, when a market order ends partially completed, etc.

edit: also we should keep in mind after merging that we still need to account for the impact on frontend / upfront sticker price for the gas

File: test/core/TakerOperations.t.sol
Comment (lnist) [reply test/core/TakerOperations.t.sol:774]: Because I set gasbase to 1 which breaks the accounting by not setting aside enough at the beginning of the market order.

File: src/MgvOfferTaking.sol
Comment (lnist) [reply src/MgvOfferTaking.sol]: Are you OK with adding this as a task to our general PR, so that this PR can be completed?

File: src/MgvOfferTaking.sol
Comment (lnist) [reply src/MgvOfferTaking.sol]: I'll resolve this comment then.

File: src/MgvOfferTaking.sol
Comment (lnist) [reply src/MgvOfferTaking.sol]: Updated in the following:

You're on the right track. The reservation in 156 is needed because 260 is an underapproximation. It goes like this when taking two offers:
1. Line 260 reached first time, we have at least `offer_gasbase+gasreq1-gasused1` gas left for rounding off the market order due to what is set aside in 156 and the accounting in internalMarketOrder. It is larger due to line 211 also conditioning on additional offer_gasbase.
2. Line 260 reached the second time we have gasreqLeft be `offer_gasbase +gasreq1-gasused1+gasreq2-gasused2`, however, note that we will not execute offer2 unless gasleft() can also cover an additional offer_gasbase when reaching line 211. 

So we will never start executing an offer unless we also have `offer_gasbase` left on top of the `gasreq` . This should always be sufficient to complete a market order of 1 offer. For many offers this code relies on offer_gasbase being able to cover for the overhead of unwrapping the stack and doing the necessary updates.
Alternatively we should add some constant overhead for each offer on line 260 or let the reservation on line 156 be another value which we know is big enough to cover the unwrap of stack.

Note: gasreq for one offer is the worst case for that offer, so with a set of offers each recursion will set aside too much in case worst case is not hit, so on average a bit extra is set aside for each offer.

File: test/core/TakerOperations.t.sol
Comment (adhusson) [reply test/core/TakerOperations.t.sol:774]: ahh of course :)

File: test/core/TheClog.t.sol
Comment (lnist) [reply test/core/TheClog.t.sol]: done

File: src/MgvGovernable.sol
Comment (espendk) [inline src/MgvGovernable.sol]: If it's safe w/o optimizer at 79, doesn't that mean that the optimizer would increase the max recursion depth? If so, wouldn't it be safe to set it to 79 as the default?

File: test/core/gas/OfferPosthookFailGasDelta.t.sol
Comment (espendk) [inline test/core/gas/OfferPosthookFailGasDelta.t.sol:69]: nifty!

File: test/lib/MangroveTest.sol
Comment (espendk) [inline test/lib/MangroveTest.sol]: use different measurement method compared to what?

File: src/MgvOfferTaking.sol
Comment (adhusson) [reply src/MgvOfferTaking.sol]: ok! (#481)

File: src/MgvGovernable.sol
Comment (lnist) [reply src/MgvGovernable.sol]: With the assumption that I haven't overlooked some edge case that is more expensive than the testet, then yes, I would expect so. And we can decrease it runtime if needed.
What do you think? Both you and @adhusson : Do you think it should be this hardcoded value or a deployment parameter?

File: test/lib/MangroveTest.sol
Comment (lnist) [reply test/lib/MangroveTest.sol]: We currently use transfers inside ActivateSemibook - IMHO this is wrong, we should use the estimate given by GasBaseTest.
(Even gasBaseTest is not the worst case, that would be with a failing posthook, but since failing posthooks does not happen every single time it may be good).

File: src/MgvGovernable.sol
Comment (espendk) [reply src/MgvGovernable.sol]: It may change over time with new EIPs, right? That could be a reason for having it a config value.
But that's probably rare, so not sure that consideration should weigh much in the decision.

My intuition is that a deployment param would be good. This will make it easy to vary depending on chain, but w/o the overhead of having it as a rarely (I expect) changed config parameter.

File: test/lib/MangroveTest.sol
Comment (espendk) [reply test/lib/MangroveTest.sol]: Ok. Did you run the GasBaseTest and set the parameter based on that?

Comment (lnist) [reply test/lib/MangroveTest.sol]: Yes, but unoptimized

Comment (espendk) [reply test/lib/MangroveTest.sol]: Ok, makes sense.

So the FIXME is for having this discussion, right? Or is there more to it?

Maybe it could be changed into a comment describing how to check/update `gasbase` for `MangroveTest`?

File: src/MgvGovernable.sol
Comment (lnist) [reply src/MgvGovernable.sol]: Constant vs config vs deployment also applies to the 7000 constant for recursion overhead

File: src/MgvOfferTaking.sol
Comment (espendk) [inline src/MgvOfferTaking.sol]:
```suggestion
      /* We have an upper limit on total gasreq for failing offers to avoid failing offers delivering nothing and exhausting gaslimit for the transaction. */
```

File: src/MgvOfferTaking.sol
Comment (espendk) [inline src/MgvOfferTaking.sol]: This can be inlined now, right?

File: src/MgvOfferTaking.sol
Comment (espendk) [inline src/MgvOfferTaking.sol]: `MgvReader` should be updated to reflect this.

File: preprocessing/structs.ts
Comment (espendk) [inline preprocessing/structs.ts]:
```suggestion
      /* * `maxGasreqForFailingOffers` is the maximum gasreq failing offers can consume in total. This is used in a protection against failing offers consuming gaslimit for transaction. Setting it too high would make it possible for successive failing offers to consume gaslimit, setting it too low will make a non-healthy book not execute enough offers. `gasmax` and `maxRecursionDepth` bit sizes constrain this.  */
```

File: src/MgvOfferTaking.sol
Comment (lnist) [reply src/MgvOfferTaking.sol]: Yes - I think that should be a separate PR

File: src/MgvOfferTaking.sol
Comment (lnist) [reply src/MgvOfferTaking.sol]: How much of it? "MAxGasreqForFailingOffers" for instance does not make sense since MgvReader assumes success, right?

Comment (espendk) [reply src/MgvOfferTaking.sol]: Indeed it assumes success. 

The MgvReader implementation closely mirrors the MgvOfferTaking implementation for ease of validation and maintenance; This PR changes the structure of the code and I think MgvReader should be updated to reflect that change.

File: test/lib/MangroveTest.sol
Comment (lnist) [reply test/lib/MangroveTest.sol]: I have updated it with a newly measured value. Note that this is for an entire market order of 1 offer to an empty book. Previously we used only transfers as baseline which would give a value of 128176.

File: src/MgvOfferTaking.sol
Comment (lnist) [reply src/MgvOfferTaking.sol]: Resolving obsolete discussion.

File: src/MgvGovernable.sol
Comment (lnist) [reply src/MgvGovernable.sol]: Resolving obsolete discussion.

File: src/MgvOfferTaking.sol
Comment (lnist) [reply src/MgvOfferTaking.sol]: Done

File: script/periphery/CopyOpenSemibooks.s.sol
Comment (espendk) [inline script/periphery/CopyOpenSemibooks.s.sol:7]: Should this be deleted?

```suggestion
```

File: script/periphery/UpdateMarket.s.sol
Comment (espendk) [inline script/periphery/UpdateMarket.s.sol:7]: Delete?

```suggestion
```

File: src/MgvOfferTakingWithPermit.sol
Comment (espendk) [inline src/MgvOfferTakingWithPermit.sol:10]: Delete?

```suggestion
```

**Review by espendk — COMMENTED** [review]
I thought there would be changes to `MgvReader`? Cf this comment: https://github.com/mangrovedao/mangrove-core/pull/461#discussion_r1315884201

**Comment by espendk** [comment]
> I thought there would be changes to `MgvReader`? Cf this comment: [#461 (comment)](https://github.com/mangrovedao/mangrove-core/pull/461#discussion_r1315884201)

Doh, there is! Somehow GitHub didn't list them in the changes since my last review...
Sorry!

File: script/periphery/CopyOpenSemibooks.s.sol
Comment (lnist) [reply script/periphery/CopyOpenSemibooks.s.sol:7]: No - they implicitly got console from reader, and I removed the import in reader, making these scripts fail. We could have imported only console, but the debug toString stuff is nice to have.

File: src/AbstractMangrove.sol
Comment (adhusson) [inline src/AbstractMangrove.sol]: I'm wondering if it shouldn't be an immutable. 
- I don't think different chains can yield different max recursion depths.
- Nor can an evolution of the gas pricing yield a different max recursion depths.
- So if the value is correct at deployment, it's correct forever (?)

Also, as long as there's a test that checks the max depth can be reached, having a wrong value there is wasteful at worst. And if we have one more test that checks (max depth + 1) cannot be reached, we canot make a wrong deployment :)

File: src/AbstractMangrove.sol
Comment (adhusson) [inline src/AbstractMangrove.sol]: Same here I would have the `10` in Constants.sol

File: src/IMangrove.sol
Comment (adhusson) [inline src/IMangrove.sol]: Following the above, I don't think `maxRecursionDepth` should be a param -- and even in a world where it's still alterable by governance, I don't see any good reason for callers to customize the value?

File: src/MgvOfferTaking.sol
Comment (adhusson) [inline src/MgvOfferTaking.sol:160]: I'm missing why we have to copy `maxGasreqForFailingOffers` from `sor` to `mor`

**Review by adhusson — CHANGES_REQUESTED** [review]
I wouldn't want to be a plumber, this is hard :D 

On top of my comments, I think we should do the "move `gasreq` to `offer` so that the code for `internalMarketOrder` doesn't have to be restructured so much with an additional `endInternalMarketOrder` function." -- I think we would want to do it anyway but doing it now prevents another PR cycle where we swap the `internalMarketOrder` structure back to where it was. (I can add it as a commit if you want)

File: test/lib/MangroveTest.sol
Comment (lnist) [reply test/lib/MangroveTest.sol]: Added FIXME in ActivateSemibook

File: src/AbstractMangrove.sol
Comment (lnist) [reply src/AbstractMangrove.sol]: As long as we hit the very worst code paths wrt stack depth, then yes it is a constant (modulo optimization). For the current tests, it fails above depth 80 with optimization and 200 runs. So setting to a constant of 75. This should leave enough room for some weird extra costly path. And in that case the path will most likely not be hit often, but if it is, we can pass in an overridden value via external calls.

File: src/IMangrove.sol
Comment (lnist) [reply src/IMangrove.sol]: It depends on how cautious  we are. If we are confident we cannot hit the 75 limit, then we can remove the override. Remember the limit only really affects successful offers since failing offers can also be controlled by maxGasreqForFailingOffers.

File: src/MgvOfferTaking.sol
Comment (lnist) [reply src/MgvOfferTaking.sol:160]: I don't understand? Maybe the code changed?

File: src/MgvGovernable.sol
Comment (lnist) [reply src/MgvGovernable.sol]: It should only change with optimizations, so we decided to set it at an expected safe 75 constant value.

File: src/MgvOfferTaking.sol
Comment (lnist) [reply src/MgvOfferTaking.sol:160]: ah, it is because I allow it to be overridden by parameter to function.

File: src/AbstractMangrove.sol
Comment (lnist) [reply src/AbstractMangrove.sol]: I have made it a deployment parameter (last commit - can revert if you disagree). Why? Because 10 was quite arbitrary. We could consider:
1) Let it be hardcoded, but with 10 in constants like you suggest, since it can be set via governance
2) Let it be a multiplier of gasmax and thus save a lot of bits in global struct.
wdyt?

Comment (adhusson) [reply src/AbstractMangrove.sol]: Hardcoded multiplier of gasmax seems like the easiest thing!

File: src/IMangrove.sol
Comment (adhusson) [reply src/IMangrove.sol]: I think that even if the 75 limit is wrong, it's fine if governance changes it, right? (assuming we don't make it an immutable)

I think that `maxRecursionDepth` works differently from `maxGasreqForFailingOffers`, which has to be adjusted depending on type of clog. Whereas `maxRecursionDepth` ideally should just be set to the perfect value and never changed?

File: src/MgvOfferTaking.sol
Comment (adhusson) [reply src/MgvOfferTaking.sol:160]: Ah yes makes sense :)

**Pull request merged**
